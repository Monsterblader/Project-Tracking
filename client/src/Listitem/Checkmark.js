import React from 'react';

function Checkmark(props) {
  return (
    <svg onClick={ props.activateCheckmark }
        height="30.0000000"
        width="30.0000000"
        transform="scale(0.75) translate(0,12.5)"
        xmlns="http://www.w3.org/2000/svg">
      <path
          d="M 5.0,18.33 L 11.66,25.0 L 25.0,8.33"
          style={{
            fill: "none",
            stroke: "#008000",
            "strokeWidth": 5.0000000,
            "strokeLinecap": "round",
            "strokeLinejoin": "round",
            "strokeDasharray": "none"
          }}
      />
    </svg>
  );
}

export default Checkmark;
