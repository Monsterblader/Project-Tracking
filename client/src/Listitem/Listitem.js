import React, { Component } from 'react';
import Checkmark from '../assets/Checkmark.js';
import Xmark from '../assets/Xmark.js';
import Edititem from './Edititem/Edititem.js';
import './Listitem.css';

class Listitem extends Component {
  constructor(props) {
    super();

    this.state = {
      edit: false,
      editedItem: props.data.item,
    };
  }

  render() {
    return (
      <div
          className="list-item"
          key={ this.props.data.id }>
        { this.state.edit
          ? <Edititem
                itemState={ this.state }
                cancel={ () => this.setState({ edit: false }) }
                changeItem={ newItem => this.setState({ editedItem: newItem }) }
                updateItem={ newItem => {
                  this.setState({ edit: false });
                  this.props.updateItem({ item: newItem });
                }}/>
          : <div>
              <div
                  className="svg-container"
                  onClick={ () => this.props.updateItem({ completed: true }) }>
                <Checkmark />
              </div>
              <div
                  className="svg-container"
                  onClick={ () => this.props.delete(this.props.data.id) }>
                <Xmark />
              </div>
              <div
                  className="message text-box"
                  onMouseOver={ () => this.setState({ edit: true })}>
                { this.state.editedItem }
              </div>
            </div>
        }
      </div>
    );
  }
}

export default Listitem;
