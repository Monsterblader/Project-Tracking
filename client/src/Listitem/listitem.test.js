// https://gist.github.com/wahengchang/108ca55981f6600c252ad0cb9d4c518f
import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import Renderer from 'react-test-renderer';
import ReactTestUtils from 'react-dom/test-utils';
import Listitem from './Listitem.js';

jest.mock('./Edititem/Edititem.js',
  () =>
    () => <div id="mock-Edititem">Mock Edititem</div>
)
jest.mock('../assets/Checkmark.js',
  () =>
    () => <div id="mock-Checkmark">Checkmark</div>
)
jest.mock('../assets/Xmark.js',
  () =>
    () => <div id="mock-Xmark">Xmark</div>
)

describe('Listitem', () => {
  const data = {
    id: 1234,
    item: "test value",
  };
  const del = jest.fn();
  const updateItem = jest.fn();
  let container = null;

  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  test('it renders "Listitem"', () => {
    const component = Renderer.create(
      <Listitem data={ data } />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders without crashing', () => {
    render(<Listitem data={ data } />, container);
  });

  it('creates an item with "test value"', () => {
    render(<Listitem data={ data } />, container);
    const el = container.getElementsByClassName('message text-box')[0];
    expect(el.textContent).toEqual(data.item);
  });

  describe('the <Checkmark />', () => {
    it('renders <Checkmark />', () => {
      render(<Listitem data={ data } />, container);
      expect(document.getElementById('mock-Checkmark')).toBeTruthy();
    });

    it('calls updateItem on click', () => {
      render(<Listitem data={ data } updateItem={ updateItem }/>, container);
      const el = document.getElementById('mock-Checkmark');
      el.parentNode.click();
      expect(updateItem).toHaveBeenCalledTimes(1);
    });
  });

  describe('the <Xmark />', () => {
    it('renders <Xmark />', () => {
      render(<Listitem data={ data } />, container);
      expect(document.getElementById('mock-Xmark')).toBeTruthy();
    });

    it('calls delete on click', () => {
      render(<Listitem data={ data } delete={ del }/>, container);
      const el = document.getElementById('mock-Xmark');
      el.parentNode.click();
      expect(del).toHaveBeenCalledWith(1234);
    });
  });

  it('renders <Listitem> when clicked', () => {
    render(<Listitem data={ data } />, container);
    const el = document.getElementsByClassName('message text-box')[0];
    el.click();
    expect(document.getElementById('mock-Edititem')).toBeTruthy();
  });
});
