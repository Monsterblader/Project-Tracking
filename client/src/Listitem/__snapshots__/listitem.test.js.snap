// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listitem it renders "Listitem" 1`] = `
<div
  className="list-item"
>
  <div>
    <div
      className="svg-container"
      onClick={[Function]}
    >
      <div
        id="mock-Checkmark"
      >
        Checkmark
      </div>
    </div>
    <div
      className="svg-container"
      onClick={[Function]}
    >
      <div
        id="mock-Xmark"
      >
        Xmark
      </div>
    </div>
    <div
      className="message text-box"
      onClick={[Function]}
    >
      test value
    </div>
  </div>
</div>
`;
